crates/ruff_db/src/source.rs

```diff
@@ -1,74 +1,139 @@
+                Notebook::empty()
+            });
+
+            return SourceFile {
+                inner: Arc::new(SourceFileInner {
+                    kind: SourceFileKind::Notebook(notebook),
+                    count: Count::new(),
+                }),
+            };
+        }
+    };
 
     let content = file.read_to_string(db);
 
-    SourceText {
-        inner: Arc::from(content),
-        count: Count::new(),
+    SourceFile {
+        inner: Arc::new(SourceFileInner {
+            kind: SourceFileKind::Text(content),
+            count: Count::new(),
+        }),
     }
 }
 
-/// Computes the [`LineIndex`] for `file`.
-#[salsa::tracked]
-pub fn line_index(db: &dyn Db, file: File) -> LineIndex {
-    let _span = tracing::trace_span!("line_index", file = ?file.debug(db)).entered();
-
-    let source = source_text(db, file);
-
-    LineIndex::from_source_text(&source)
-}
-
-/// The source text of a [`File`].
+/// A source file that can be a text file or a notebook.
 ///
 /// Cheap cloneable in `O(1)`.
 #[derive(Clone, Eq, PartialEq)]
-pub struct SourceText {
-    inner: Arc<str>,
-    count: Count<Self>,
+pub struct SourceFile {
+    inner: Arc<SourceFileInner>,
 }
 
-impl SourceText {
-    pub fn as_str(&self) -> &str {
-        &self.inner
+impl SourceFile {
+    /// Returns the python code as a `str`.
+    pub fn code(&self) -> &str {
+        match &self.inner.kind {
+            SourceFileKind::Text(source) => source,
+            SourceFileKind::Notebook(notebook) => notebook.source_code(),
+        }
+    }
+
+    /// Returns the underlying notebook if this is a notebook file.
+    pub fn as_notebook(&self) -> Option<&Notebook> {
+        match &self.inner.kind {
+            SourceFileKind::Notebook(notebook) => Some(notebook),
+            SourceFileKind::Text(_) => None,
+        }
+    }
+
+    /// Returns `true` if this is a notebook source file.
+    pub fn is_notebook(&self) -> bool {
+        matches!(&self.inner.kind, SourceFileKind::Notebook(_))
     }
 }
 
-impl Deref for SourceText {
+impl Deref for SourceFile {
     type Target = str;
 
     fn deref(&self) -> &str {
-        self.as_str()
+        self.code()
     }
 }
 
-impl std::fmt::Debug for SourceText {
+impl std::fmt::Debug for SourceFile {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.debug_tuple("SourceText").field(&self.inner).finish()
+        let mut dbg = f.debug_tuple("SourceFile");
+
+        match &self.inner.kind {
+            SourceFileKind::Text(text) => {
+                dbg.field(text);
+            }
+            SourceFileKind::Notebook(notebook) => {
+                dbg.field(notebook);
+            }
+        }
+
+        dbg.finish()
     }
 }
 
+#[derive(Eq, PartialEq)]
+struct SourceFileInner {
+    count: Count<SourceFile>,
+    kind: SourceFileKind,
+}
+
+#[derive(Eq, PartialEq)]
+enum SourceFileKind {
+    Text(String),
+    Notebook(Notebook),
+}
```

Comment:
Ignore if this isn't relevant but I was wondering if there's any reason for not using the `SourceKind` or borrowing the name:

https://github.com/astral-sh/ruff/blob/9a2dafb43d36b62d45a604dea58224a0884cd6e5/crates/ruff_linter/src/source_kind.rs#L19-L25

Reply:
I wasn't aware of it. There are too many `PySourceType` and kind types haha. 

The only reason right now is that it's in `ruff_linter` and we can't provide all it's API (e.g. we don't want `from_path`)